File: sdks/rust/src/nodes/transform_data.rs
use async_trait::async_trait;
use std::collections::HashMap;

use crate::errors::NodeError;
use crate::node::NodeHandler;
use crate::types::Context;

/// TransformDataNode transforms JSON data based on field mappings.
///
/// Config:
///   - `mappings` (object): Map of target field name → source field path (dot-notation)
///   - `include_only` (array, optional): Only include these fields
///   - `exclude` (array, optional): Exclude these fields
///   - `defaults` (object, optional): Default values for missing fields
pub struct TransformDataNode;

#[async_trait]
impl NodeHandler for TransformDataNode {
    async fn execute(
        &self,
        ctx: &mut Context,
        config: &HashMap<String, serde_json::Value>,
    ) -> Result<serde_json::Value, Box<dyn std::error::Error + Send + Sync>> {
        let body = ctx
            .request
            .body
            .as_object()
            .ok_or_else(|| NodeError::validation("request body must be a JSON object"))?;

        let mut result = serde_json::Map::new();

        // Apply field mappings if configured
        if let Some(mappings) = config.get("mappings").and_then(|v| v.as_object()) {
            for (target, source_path) in mappings {
                if let Some(source) = source_path.as_str() {
                    if let Some(value) = get_nested_value(&ctx.request.body, source) {
                        result.insert(target.clone(), value.clone());
                    }
                }
            }
        } else {
            // No mappings — copy all fields
            for (k, v) in body {
                result.insert(k.clone(), v.clone());
            }
        }

        // Apply include_only filter
        if let Some(include_only) = config.get("include_only").and_then(|v| v.as_array()) {
            let allowed: Vec<&str> = include_only.iter().filter_map(|v| v.as_str()).collect();
            result.retain(|k, _| allowed.contains(&k.as_str()));
        }

        // Apply exclude filter
        if let Some(exclude) = config.get("exclude").and_then(|v| v.as_array()) {
            for field in exclude.iter().filter_map(|v| v.as_str()) {
                result.remove(field);
            }
        }

        // Apply defaults
        if let Some(defaults) = config.get("defaults").and_then(|v| v.as_object()) {
            for (k, v) in defaults {
                if !result.contains_key(k) {
                    result.insert(k.clone(), v.clone());
                }
            }
        }

        let output = serde_json::Value::Object(result);
        ctx.set_var("transformed_data", output.clone());

        Ok(output)
    }
}

fn get_nested_value<'a>(data: &'a serde_json::Value, path: &str) -> Option<&'a serde_json::Value> {
    let mut current = data;
    for part in path.split('.') {
        current = current.get(part)?;
    }
    Some(current)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testing::MockContext;

    #[tokio::test]
    async fn test_transform_copy_all() {
        let node = TransformDataNode;
        let mut ctx = MockContext::new()
            .with_body(serde_json::json!({"name": "John", "age": 30}))
            .build();
        let result = node.execute(&mut ctx, &HashMap::new()).await.unwrap();
        assert_eq!(result["name"], "John");
        assert_eq!(result["age"], 30);
    }

    #[tokio::test]
    async fn test_transform_with_mappings() {
        let node = TransformDataNode;
        let mut ctx = MockContext::new()
            .with_body(serde_json::json!({"first_name": "John", "last_name": "Doe"}))
            .build();
        let config = HashMap::from([(
            "mappings".into(),
            serde_json::json!({"full_name": "first_name", "surname": "last_name"}),
        )]);
        let result = node.execute(&mut ctx, &config).await.unwrap();
        assert_eq!(result["full_name"], "John");
        assert_eq!(result["surname"], "Doe");
    }

    #[tokio::test]
    async fn test_transform_with_exclude() {
        let node = TransformDataNode;
        let mut ctx = MockContext::new()
            .with_body(serde_json::json!({"name": "John", "secret": "hidden", "age": 30}))
            .build();
        let config = HashMap::from([("exclude".into(), serde_json::json!(["secret"]))]);
        let result = node.execute(&mut ctx, &config).await.unwrap();
        assert!(result.get("secret").is_none());
        assert_eq!(result["name"], "John");
    }

    #[tokio::test]
    async fn test_transform_with_defaults() {
        let node = TransformDataNode;
        let mut ctx = MockContext::new()
            .with_body(serde_json::json!({"name": "John"}))
            .build();
        let config = HashMap::from([(
            "defaults".into(),
            serde_json::json!({"role": "user", "active": true}),
        )]);
        let result = node.execute(&mut ctx, &config).await.unwrap();
        assert_eq!(result["role"], "user");
        assert_eq!(result["active"], true);
    }
}

File: sdks/go/examples/nodes/transformdata.go
package nodes

import (
	"fmt"
	"strings"

	blok "github.com/nickincloud/blok-go"
)

// TransformDataNode transforms JSON data based on field mappings.
//
// Config:
//   - mappings (map[string]string): Field mapping from source to target.
//     Keys are target field names, values are source field paths (dot-notation).
//   - include_only ([]string, optional): If set, only include these fields from input.
//   - exclude ([]string, optional): Fields to exclude from output.
//   - defaults (map[string]any, optional): Default values for missing fields.
//
// Input (request body):
//   - The input data object to transform
//
// Output:
//   - The transformed data object
type TransformDataNode struct{}

// Execute performs the data transformation.
func (n *TransformDataNode) Execute(ctx *blok.Context, config map[string]interface{}) (interface{}, error) {
	body := ctx.Request.BodyMap()
	if body == nil {
		return nil, blok.NewValidationError("request body must be a JSON object")
	}

	result := make(map[string]interface{})

	// Apply field mappings if configured
	if mappings, ok := config["mappings"].(map[string]interface{}); ok {
		for targetField, sourcePathRaw := range mappings {
			sourcePath, ok := sourcePathRaw.(string)
			if !ok {
				continue
			}
			value := getNestedValue(body, sourcePath)
			if value != nil {
				result[targetField] = value
			}
		}
	} else {
		// No mappings — copy all fields
		for k, v := range body {
			result[k] = v
		}
	}

	// Apply include_only filter
	if includeOnly, ok := config["include_only"].([]interface{}); ok && len(includeOnly) > 0 {
		filtered := make(map[string]interface{})
		for _, field := range includeOnly {
			fieldName, ok := field.(string)
			if !ok {
				continue
			}
			if v, exists := result[fieldName]; exists {
				filtered[fieldName] = v
			}
		}
		result = filtered
	}

	// Apply exclude filter
	if exclude, ok := config["exclude"].([]interface{}); ok {
		for _, field := range exclude {
			fieldName, ok := field.(string)
			if !ok {
				continue
			}
			delete(result, fieldName)
		}
	}

	// Apply defaults for missing fields
	if defaults, ok := config["defaults"].(map[string]interface{}); ok {
		for k, v := range defaults {
			if _, exists := result[k]; !exists {
				result[k] = v
			}
		}
	}

	// Store transformed data in vars
	ctx.SetVar("transformed_data", result)

	return result, nil
}

// getNestedValue retrieves a value from a nested map using dot-notation path.
func getNestedValue(data map[string]interface{}, path string) interface{} {
	parts := strings.Split(path, ".")
	var current interface{} = data

	for _, part := range parts {
		m, ok := current.(map[string]interface{})
		if !ok {
			return nil
		}
		current, ok = m[part]
		if !ok {
			return nil
		}
	}

	return current
}

// InputSchema returns the JSON Schema for validation.
func (n *TransformDataNode) InputSchema() map[string]interface{} {
	return map[string]interface{}{
		"type": "object",
	}
}

// OutputSchema returns the JSON Schema for validation.
func (n *TransformDataNode) OutputSchema() map[string]interface{} {
	return map[string]interface{}{
		"type": "object",
	}
}

// Ensure TransformDataNode implements ValidatedNodeHandler
var _ blok.ValidatedNodeHandler = (*TransformDataNode)(nil)

// compile-time interface check for other nodes
var _ blok.NodeHandler = (*HelloWorldNode)(nil)
var _ blok.NodeHandler = (*ApiCallNode)(nil)

// NodeName constants for registration.
const (
	HelloWorldNodeName    = "hello-world"
	ApiCallNodeName       = "api-call"
	TransformDataNodeName = "transform-data"
	ChainTestNodeName     = "chain-test"
)

// RegisterAll registers all example nodes with the registry.
func RegisterAll(registry *blok.NodeRegistry) {
	registry.Register(HelloWorldNodeName, &HelloWorldNode{})
	registry.Register(ApiCallNodeName, &ApiCallNode{})
	registry.Register(TransformDataNodeName, &TransformDataNode{})
	registry.Register(ChainTestNodeName, &ChainTestNode{})
}

// RegisterDefaults is an alias for RegisterAll for convenience.
func RegisterDefaults(registry *blok.NodeRegistry) {
	RegisterAll(registry)
}

// Describe returns a description of all example nodes.
func Describe() map[string]string {
	return map[string]string{
		HelloWorldNodeName:    "Greets the user with a configurable prefix and name",
		ApiCallNodeName:       "Makes HTTP requests to external APIs",
		TransformDataNodeName: "Transforms JSON data based on field mappings",
	}
}

// DescribeNode returns a description of a specific example node.
func DescribeNode(name string) (string, error) {
	descriptions := Describe()
	desc, ok := descriptions[name]
	if !ok {
		return "", fmt.Errorf("unknown example node: %s", name)
	}
	return desc, nil
}
